### PR TITLE
Extract signup analytics to a separate module

### PR DESCRIPTION
**Pull request by jsnajdr**
A small step to make `lib/analytics` smaller and more focused: move all functions that are specific to signup (record signup start and completion) out of `lib/analytics` to a separate module.

Also refactors the signup ad-tracking functions to use `adTrackSomething` naming convention (prevents naming conflicts when two modules export different functions with the same `recordSomething` name).

And merges `recordSignupCompletionInFloodlight` and `recordSignup` into one function `adTrackSignupComplete`.

### CODE REVIEW

**Review by andrewserong — APPROVED** [review]
This is looking so much cleaner, thanks for the refactor @jsnajdr and the clear re-naming of some of these functions — tested the Tracks and analytics calls across normal signup, passwordless, and social login, each of the steps, and it's all working correctly.

I noticed that this PR also removes the forked flow as in #38138, personally I think this one is fine to remove now, but just CCing in @michaeldcain as well on this one just in case.

The only other complexity I can think of is this open PR (https://github.com/Automattic/wp-calypso/pull/37410) which will need to be refactored after this one lands. CC @MicBosi

**Comment by jsnajdr** [comment]
> I noticed that this PR also removes the forked flow as in #38138

Yes, I intentionally extracted that bit of code into a separate PR, as it's a different kind of change rather than mere refactoring.

> The only other complexity I can think of is this open PR (#37410)

Yes, this PR will indeed throw a wrench into that work 🙂 I can help you and @MicBosi to rebase and resolve the conflicts.

**Review by MicBosi — CHANGES_REQUESTED** [review]
@jsnajdr I have a counter proposal. Instead of splitting by type of events (sign up in this particular instance, but there are at least other 4 main events atm), we could consider a different axis of separation. 

One example, which mimics the direction we've tried to go could be to have 1 Facade object (more or less represented by the current `analytics` object) + multiple analytics "engines" (or an equivalent Subject/Observer pattern). Example:

- 1 Analytics wrapper class similar to what we have now with it's 5 main methods/events (registration, signup, add to cart, purchase, page view).

- N analytics "engines" that provide a concrete implementation for the 5 main events (plus their idiosyncratic features that might generalize beyond the 5 events), ex. Google Analytics, Tracks, Ad-Tracking, Google Tag Manager etc.

There is a whole spectrum of possibilities from more practical to extremely metaphysical (I usually tend to thread on the practical side). One could even argue that the current `analytics` object currently acts as an hub that could be substituted with one of the event-dispatching mechanisms that already exist in Calypso and we only need to turn the various Tracks, Ad-Tracking etc. into listeners of specific events.

Either way the first step in this direction would seem to be moving the Tracks, Ad-Tracking, etc. logic into their own modules.

cc @jaswrks

**Comment by jsnajdr** [comment]
Hi @micbosi, thanks for the feedback and the proposal!

The ultimate goal for `lib/analytics` is to publish it as an independent npm package, `@automattic/analytics`, and use it at multiple places, e.g., Calypso, Jetpack, Woo, and various WP.com code bases outside Calypso like Full Site Editing or Gutenboarding.

An important design goal for a JS library running in browser is good support for code splitting. If a project doesn't use, say, Google Analytics or Facebook Ad Tracking, the JS bundle for that project shouldn't contain that code at all -- it should be optimized out by webpack and other tools in the build chain.

Similarly, the generic `@automattic/analytics` shouldn't contain analytics code for Calypso Signup or Checkout. Not only they are not interesting at all outside Calypso, but they also shouldn't be bundled in the Jetpack JS bundle, for example. That's why I'm trying to move out the section specific code from `lib/analytics` to better places like `client/signup` or `client/checkout`.

Having one giant `analytics` object with dozens of methods doesn't support code splitting very well. There is a lot of funny ways how one can call an object method:
```js
// the normal way
analytics.recordPurchase();

// computed property access
analytics[ 'recordPurchase' ]();

// a nontrivially computed property
const method = 'recordCart' + adding ? 'Add' : 'Remove';
analytics[ method ]();

// there are more crazy ways
const method = Object.keys( analytics ).find( name => /Purchase$/.test( name ) );
method.call( analytics );
```

In short, the build tools cannot reliably figure out if a method is called or not, and cannot optimize it out. That's why the "Facade Object" approach is not a good route to take.

Exporting functions from ES modules and `import`-ing them at other places works much better: imports and exports can be statically analyzed and the bundler can be 100% sure that a certain function is never called and remove it from the output.

I'd like to gradually split the big `lib/analytics` module (mainly `index.js` and `ad-tracking.js`) into independent smallish modules and import them only when needed. And remove the object-oriented `analytics.tracks.recordEvent` convention, because it's not a good design approach for browser JS libraries.

Does that make sense?

**Comment by MicBosi** [comment]
Thanks for the explanation @jsnajdr. 

> The ultimate goal for lib/analytics is to publish it as an independent npm package, @automattic/analytics, and use it at multiple places, e.g., Calypso, Jetpack, Woo, and various WP.com code bases outside Calypso like Full Site Editing or Gutenboarding.

This might be a matter of syntax but how about considering the current lib/analytics a Calypso-Marketing analytics library (which I think it's fair to say it's what it is) and why not start from a clean slate for the generic analytics library so that it's independent from the current library's quirks.

Even if right now we had such a generic analytics library, we'd still need in some form a Calypso library or set of modules that coordinates the various Calypso-specific needs. That is basically what we have now, although in a sub-optimal meshed-together form.

I agree using JS proper objects does not allow for easy static analysis, but we can address that by simply turning each of those objects into their own modules that export their functions.

The entire `analytics` can be it's own module. That would probably be the last one to be modularized though as it's more practical to start from the leaves for the dependency tree and work our way up, ie staring the modularization from the `analytics.tracks` object, `analytics.ga` object, `ad-tracking.js` could go in its own library, same for `utils` etc.

> Similarly, the generic @automattic/analytics shouldn't contain analytics code for Calypso Signup or Checkout. Not only they are not interesting at all outside Calypso, but they also shouldn't be bundled in the Jetpack JS bundle, for example. That's why I'm trying to move out the section specific code from lib/analytics to better places like client/signup or client/checkout.

I agree 100%. Which is why I advocate starting such a library from a clean slate and in parallel work on the modularization of the Calypso analytics library, which rightly focuses on Calypso's needs.

> I'd like to gradually split the big lib/analytics module (mainly index.js and ad-tracking.js) into independent smallish modules and import them only when needed. And remove the object-oriented analytics.tracks.recordEvent convention, because it's not a good design approach for browser JS libraries.

I 100% support this effort. The only difference is that I would prefer to see that broken down logic living into their own `analytics-dispatcher`, `analytics-tracks`, `analytics-ad-tracking` etc. modules, rather than having each event type being handled in a different part of Calypso.

Ultimately the `signup` module or `checkout` module, should not be concerned on whether `identifyUser()` comes before or after registration or whether Google Analytics should be called before or after Tracks and handle the subtle differences that sometimes might arise between them. Which is why I'm a big fan of restructuring the current analytics library in modules or even sub-libraries but not of spreading the analytics concerns across Calypso's code base into non-analytics related modules.

Another reasonable assumption could also be that once we have those modules ironed out, we will also have a cleaner view of what should or could be generalized and gradually moved out into the generic analytics library.

On the other hand, if there is already a design in place for the new generic analytics library let's start implement that in a clean sterile place and gradually bring it in Calypso. That could also work.

File: client/lib/signup/analytics.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+		// Delay using the analytics localStorage queue.
+		return analytics.queue.add(
+			'recordSignupComplete',
+			{ flow, isNewUser, isNewSite, hasCartItems, isNew7DUserSite },
+			true
+		);
+	}
+
+	// Tracks
+	analytics.tracks.recordEvent( 'calypso_signup_complete', {
```
Comment (jaswrks) [inline client/lib/signup/analytics.js]: @jsnajdr Thanks for tackling this. The improvements are much needed and your work is certainly appreciated. I also appreciate that you're involving marketing devs in this decision. Really nice, thanks!

You wrote...

> I'm trying to move out the section specific code from lib/analytics
> to better places like client/signup or client/checkout.

@MicBosi  writes...

> I'm a big fan of restructuring the current analytics library in modules or even sub-libraries but not of spreading the analytics concerns across Calypso's code base into non-analytics related modules.

I just want to add a ➕**1** on this because the `calypso_signup_start` and `calypso_signup_complete` events, among others (@MicBosi can provide a more complete list) are in the critical path where paid search, display ads, and Google Optimize A/B tests all get their data. Sometimes using Tracks, sometimes using GA events, sometimes using ad tracking. As a result, Marketing devs are frequently asked to audit, modify, or otherwise update these.

If we can find a way to keep all of mission-critical events in one place (maybe we group them together and make these an exception) that would be great. Ideally, the Tracks, GA, and ad tracking calls are all together so that we can be absolutely sure that changing props or adding something new is going to cover us in all cases.

What would be bad is if we changed the props in this block of code, and only later we discover there was another instance of this `calypso_signup_complete` event being fired. Or, as happened recently, a suffixed variation.

What do you think?

Comment (jaswrks) [reply client/lib/signup/analytics.js]: Also, I realize searching the codebase is easy enough when one knows the event they're looking for. I'm thinking about onboarding & discoverability for marketing devs when I imagine these events being together in one place.

**Comment by jsnajdr** [comment]
Thanks @MicBosi and @jaswrks for the great feedback!

> This might be a matter of syntax but how about considering the current lib/analytics a Calypso-Marketing analytics library (which I think it's fair to say it's what it is) and why not start from a clean slate for the generic analytics library so that it's independent from the current library's quirks.

I think it's also a matter of personal preference and working style when choosing one of these approaches. For me, being able to refactor a part of the library and move it into a separate module means that I understand the library better than before. If I'm able to change it, it means I understand it! At this moment I feel I'm still in the learning phase, and yet I'm able to ship tangible improvements.

Creating an empty `@automattic/analytics` package and start putting code into it from scratch is possible, but I'd prefer to land a few more PRs like this one before starting with that.

> Which is why I'm a big fan of restructuring the current analytics library in modules or even sub-libraries but not of spreading the analytics concerns across Calypso's code base into non-analytics related modules.

I think this is the main concern you're raising, is that right? That the Calypso analytics code should remain concentrated in the `lib/analytics` module, even if it's section specific. The directory structure should look like this:
```
lib/analytics
  tracks.js
  google-analytics.js
  statsd.js
  facebook.js
  bing.js
  ...
  signup.js
  cart.js
  store.js
  page-view.js
```
I.e., break up the giant `index.js` and `ad-tracking.js` files into smaller modules. The first half of the above list are generic services that will eventually end up in `@automattic/analytics`. The second half is Calypso-specific and will stay in `lib/analytics`.

There shouldn't be modules like `lib/signup/analytics` or `lib/cart/analytics` that spread the analytics code all over the folder tree. Would that help?

**Comment by MicBosi** [comment]
> I think it's also a matter of personal preference and working style when choosing one of these approaches.

@jsnajdr Definitely, we all have different mental models about things, I too am constantly learning from others' perspectives.

> I think this is the main concern you're raising, is that right? That the Calypso analytics code should remain concentrated in the lib/analytics module, even if it's section specific.

> There shouldn't be modules like lib/signup/analytics or lib/cart/analytics that spread the analytics code all over the folder tree. Would that help?

Correct, I would say that is more or less the principal factor, although I don't have a strong opinion on the implementation details. Splitting the modules and keeping them in one `analytics` directory sounds good for me. But I also have nothing against having different `analytics-*` dirs if that is desirable or required for some reason. As long as they are easy to find and live "close" to one another so it's easy to maintain and keep consistency.

We followed a similar approach to our `a8c-analytics` library in WP.com (`wp-content/mu-plugins/a8c-analytics/`) which although still a fairly crude piece of code it has hugely simplified the management of all things Privacy & Analytics for all the main company's sites (WP.com, JP.com, Automattic.com etc.). Previously all that logic was spread across every site's code base, sometimes in specific themes, sometimes in random plugins etc. Now if anyone wants to update `en.support.wordpress.com` Google Analytics tag for instance knows that everything is in there and can even see how other sites handle analytics and trackers for inspiration.

---

You may be already aware of these but here's a couple of heads up with regard to the present and future of analytics with respect to Calypso.

1) The cookie banner (`client/blocks/gdpr-banner/`) is an essential part of the analytics puzzle due to the GDPR and new and coming  privacy related regulations around the world, since many of these trackers have to go through a preliminary phase of assessing whether the user has given consent or not and whether based on geo location we should ask for consent or not, before firing the various events. Something to keep in mind for any future analytics library. At the moment the cookie banner is fairly self contained consisting of mostly UI logic, the `shouldShowBanner()` might be a good candidate to be moved out of it in order to group together the decision making logic in the appropriate modules (ie `shouldShowBanner()` could probably live happily where `isCurrentUserMaybeInGdprZone()` etc. is also defined, wherever that will be).

2) We are in the process of rolling out Google Tag Manager in WordPress.com, in fact Fri next week we will have completed the roll out on every landing page, the logged-out home page, `/go` etc. @jaswrks is leading the roll out. 
In Q1 2020 we will start bringing GTM on Calypso. The two main aspects I wanted to share is that:
a) once we have GTM in place, most of the vendor specific code like Facebook, Google Ads etc. that now is making `ad-tracking.js` large will disappear as those will be handled by GTM itself
b) Google Analytics itself will also be handled within GTM, which means the `analytics.ga` object will also become redundant and disappear. Although we may still want to create some kind of proxy mechanism to allow people to easily send arbitrary events to GA which is easy now, but with GTM becomes a bit more cumbersome. 

In theory we could also move Tracks inside GTM at least to some degree but I think practical aspects will probably prevail there and we'll keep Tracks as it is given it's used much more pervasively throughout Calypso and also I personally like the idea of keeping the two tracking systems independent from one another to improve redundancy and resiliency. @jaswrks what do you think?

**Comment by jaswrks** [comment]
@jsnajdr  writes...

> There shouldn't be modules like lib/signup/analytics or lib/cart/analytics that spread the analytics code all over the folder tree. Would that help?

Yes, that would help. Thank you!

> I.e., break up the giant index.js and ad-tracking.js files into smaller modules. The first half of the above list are generic services that will eventually end up in @automattic/analytics. The second half is Calypso-specific and will stay in lib/analytics.

Perfect. 🥇 

---

@MicBosi writes...

> In theory we could also move Tracks inside GTM at least to some degree but I think practical aspects will probably prevail there and we'll keep Tracks as it is given it's used much more pervasively throughout Calypso and also I personally like the idea of keeping the two tracking systems independent from one another to improve redundancy and resiliency. @jaswrks what do you think?

I'll echo your thoughts on this. I like the plan that keeps Tracks and GTM independent from one another. I agree that it improves redundancy and resiliency. 

> Although we may still want to create some kind of proxy mechanism to allow people to easily send arbitrary events to GA which is easy now, but with GTM becomes a bit more cumbersome

Yeah, I think there are still some question marks about how we can best handle GA events in Calypso once we introduce GTM. Each of the GA events still needs to be trigged by Calypso like they are now, but instead of those going straight to Google Analytics, GTM intercepts, and then those managing GTM can decide what to do with them.

So what is currently `analytics.ga.recordEvent()` will end up, in one form or another, setting variables in the `dataLayer`, triggering a GTM event, and then GTM will take the event name and variables and handle things appropriately, as configured by us, or by marketers.

**Comment by jsnajdr** [comment]
Hi @micbosi and @jaswrks 👋 

To address the main concern, I pushed a commit (https://github.com/Automattic/wp-calypso/pull/37913/commits/f66ba0ac2f8dd585ab0bc584284b2d9fc9bfaef7) that moves several section-specific analytics modules to `lib/analytics`. Now there are:
- `lib/analytics/signup` (signup progress and user registration)
- `lib/analytics/cart` (adding and removing items from shopping cart)
- `lib/analytics/store-transactions` (paying for stuff, will plug-in into a `composite-checkout` hook soon)

That keeps the analytics concerns all concentrated in `lib/analytics`.

And thanks for sharing your future plans about Calypso analytics.

> We are in the process of rolling out Google Tag Manager in WordPress.com

If I understand GTM correctly, it will let us remove almost the whole `ad-tracking` module? Each one from the 10+ analytics libraries seems to follow the same pattern:
0. Check if the given analytics library is enabled at all: `if ( ! isFooAnalyticsEnabled ) return;`
1. Add a queue array to `window`, e.g., `window._foo = [];`
2. Record events by pushing library-specific data to that array: `window._foo.push( [ 'record', 1, 2, 3 ] );`
3. Start loading a 3rd party script from some external URL and initialize the given library with our tracking ID etc.
4. The library will take over `window._foo`, processes the accumulated queue and replaces `window._foo_lytics.push` with some function.
5. Further calls to `window._foo.push()` will invoke a function from the 3rd party script.

With GTM, we'll load just one library, Google Tag Manager, which will manage its own queue, and all the enable/disable toggles and tracking IDs are managed in the GTM web UI?

> In theory we could also move Tracks inside GTM at least to some degree but I think practical aspects will probably prevail there and we'll keep Tracks as it is

Yes, I agree that the A8c custom services, i.e., Tracks, MC stats and statsd will probably remain outside GTM. It feels a bit wrong to start relying on a third party script for that, with little benefit in return.

But Google Analytics could be managed by Google Tag Manager just fine. We already rely on Google here, so it doesn't add any additional failure point.

The fact that GA are pervasively used across Calypso shouldn't matter much. The UI code will continue to call `analytics.ga.recordEvent` or dispatch `recordGoogleEvent` Redux actions, and doesn't need to care much how exactly that call is implemented.

> many of these trackers have to go through a preliminary phase of assessing whether the user has given consent or not and whether based on geo location we should ask for consent or not, before firing the various events

That's the logic that already lives in the `refreshCountryCodeCookieGdpr`, `isCurrentUserMaybeInGdprZone` and `mayWeTrackCurrentUserGdpr` functions, is that right?

It's true that these functions don't share any code with the `GdprBanner` UI component, although they both work with the `sensitive_pixel_option` cookie.

**Comment by MicBosi** [comment]
> To address the main concern, I pushed a commit (f66ba0a) that moves several section-specific analytics modules to lib/analytics.

Thanks @jsnajdr for the great work, that all looks great to me! 👍 

> If I understand GTM correctly, it will let us remove almost the whole ad-tracking module?

> With GTM, we'll load just one library, Google Tag Manager, which will manage its own queue, and all the enable/disable toggles and tracking IDs are managed in the GTM web UI?

Correct, that is the ideal we will be pursuing 😃

**Comment by jsnajdr** [comment]
@MicBosi are there any other changes you'd like to see in this PR, or do you think it's ready to merge?

And one more question about Google Tag Manager: can it also be helpful to simplify a function like `recordOrder` in `ad-tracking.js`, which takes an info about an order and records it into 10+ services, with various parameters?

In other words, can you define a "record order" action in the GTM UI, configure the services and params there, and then record an order from Calypso with one line of code?

Skimming through GTM docs, it seems to me that's exactly what tags, triggers and the data layer are for, but I'm not sure.

**Comment by MicBosi** [comment]
> And one more question about Google Tag Manager: can it also be helpful to simplify a function like recordOrder in ad-tracking.js, which takes an info about an order and records it into 10+ services, with various parameters?

@jsnajdr Yes exactly, with GTM you push all necessary purchase info in the dataLayer which is basically a big key/value vocabulary that GTM keeps track of and that Tags can read into and then we can push a `record-order` event to GTM which we then use to fire a GTM Trigger which then can be used by N simultaneous GTM Tags (tag~=tracker in GTM) to fire Facebook, Google Ads etc. pixels.

**Comment by MicBosi** [comment]
> are there any other changes you'd like to see in this PR, or do you think it's ready to merge?

@jsnajdr everything looks 👍  to me, thanks for the great work! 

I couldn't test it just yet as I'm 150% into the finishing stages of the new homepage launch, once that is done I'll be able to test it as well, probably early next week. If you'd like to deploy asap feel free to grab anyone from the MarTech team, it would be good to have any of them reviewing the PR before lading it.

**Comment by jsnajdr** [comment]
Rebased onto latest `master` and resolved conflicts (in passwordless signup and the removal of reCAPTCHA from there).

@MicBosi is it now a good time to give this the final :eyes: and 👍 ?

**Comment by jsnajdr** [comment]
Rebased onto latest `master`, no conflicts. Still ready for review by @MicBosi and team 🙂

**Comment by MicBosi** [comment]
Hi @jsnajdr, we are just back from our meetup, I'll do my best to review by EOD. Looking forward to get this shipped! Thank you!

File: client/lib/analytics/signup.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	adTrackSignupStart( flow );
+}
+
+export function recordSignupComplete(
+	{ flow, isNewUser, isNewSite, hasCartItems, isNew7DUserSite },
+	now
+) {
+	if ( ! now ) {
+		// Delay using the analytics localStorage queue.
+		return analytics.queue.add(
```
Comment (MicBosi) [inline client/lib/analytics/signup.js]: @jsnajdr I believe the `analytics.queue` object expects the passed function name to be part of the `analytics` object and since it is not the case anymore, the signup tracking appears broken for me, ie I don't see it being fired.

For example after completing the signup process I don't see any reference to `calypso_signup_complete` in my browser's Network activity.

The initial signup part seems ok both the cookie banner and the gdpr logic seem to work as expected. To fully test all code paths one needs to be connected through a GDPR country and make sure the cookie banner appears and then click it.

This is what I used to enable all relevant features and see the logs:
```
document.cookie = 'flags=gdpr-banner,google-analytics,ad-tracking; path=/; domain=.' + document.location.hostname.split( '.' ).slice( -2 ).join( '.' );
localStorage.setItem( 'debug', 'calypso:analytics:*' );
```

**Review by MicBosi — CHANGES_REQUESTED** [review]

File: client/lib/analytics/signup.js (hunk 2 above)
Comment (jsnajdr) [reply client/lib/analytics/signup.js]: That's a great catch @MicBosi :+1: I didn't know what `analytics.queue` does until now.

I fixed this bug by importing `recordSignupComplete` from the separate module and adding a property to `analytics`. That defeats the attempt to improve code splitting, but I don't see any better solution right now.

The queueing means that `recordSignupComplete` is not called only in the Signup section, but potentially also anywhere else -- on the first page view after loading, whatever page that view is.

@jaswrks Do I understand correctly that `analytics.queue` prevents cancelling the `recordSignupComplete` async task when leaving the current page? It needs to wait for the tracker scripts to load etc.

How does Google Tag Manager handle such a case? It looks like something that's common and happens all the time: sending analytics correctly and completely before leaving a page.

Comment (MicBosi) [reply client/lib/analytics/signup.js]: @jsnajdr Thanks for fixing that.

> Do I understand correctly that analytics.queue prevents cancelling the `recordSignupComplete` async task when leaving the current page? It needs to wait for the tracker scripts to load etc.

That sounds about right to me and for the Signup Complete event in particular it's needed because we do a full reload of the page which doesn't usually happen. As far as I know waiting for the trackers to load hasn't been that big of an issue, while the page reload we do right after signup complete is.

To be honest I'm not sure how much better our tracking is with the queue mechanism, that is, it probably is but I don't have hard numbers on the topic. I also wish we didn't have to use it so if we figured out a way to not need it I think we would all be happy about it.

> How does Google Tag Manager handle such a case? It looks like something that's common and happens all the time: sending analytics correctly and completely before leaving a page.

Yes it is a common problem in general although not so much in Calypso since most page views are "virtual" page views, we don't actually reload the page, while for landing pages for example this is a real problem for things like click tracking.

In theory the Beacon API has been introduced to address this very issue but it needs to be supported by any given tracker. Google Analytics does for example. Many others don't, including Tacks. Google Tag Manager unfortunately does not affect this issue at all since supporting the Beacon API depends entirely on the tracker being fired within GTM. So from this particular point of view GTM won't make things neither better or worse as it will be triggered from within the `recordSignupComplete` function.

Comment (jsnajdr) [reply client/lib/analytics/signup.js]: > In theory the Beacon API has been introduced to address this very issue but it needs to be supported by any given tracker.

Also, the Beacon API is useful only when you want to send a single XHR request and want a guarantee that it will finish even if the page gets unloaded at the same time.

A `recordSignupComplete` call is potentially much complex than that. It may send a geolocation request to the REST API to figure out if it's in a GDPR country, then load several tracking scripts, and only then send the actual beacon tracking requests. The page can be long gone before that beacon requests even start.

In any case, @MicBosi please let me know if this PR can be merged now. #39001 depends on the changes here.

**Review by MicBosi — APPROVED** [review]
@jsnajdr LGTM! 🚢 

Everything looks good to me 👍 😃
